README.md
crypto-katz-lindell
===

copyright (c) Ben Selfridge 2021

A simple formulation of encryption schemes, based on the presentation in
*Introduction to Modern Cryptography* by Jonathan Katz and Yehuda Lindell.


CHANGELOG.md
# Revision history for crypto-katz-lindell

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Test.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}

module Main where

import Crypto.Schemes

import Control.Category
import Control.Lens (prism')
import qualified Data.BitVector.Sized as BV
import Data.Parameterized.NatRepr
import GHC.TypeNats
import Test.QuickCheck
import Test.QuickCheck.Monadic
import Test.Tasty
import Test.Tasty.QuickCheck

import Prelude hiding (id, (.))

newtype KeyLength = KeyLength Int
  deriving (Eq, Show)

instance Arbitrary KeyLength where
  arbitrary = KeyLength <$> chooseInt (1, 20)

newtype ABV w = ABV { unABV :: BV.BV w }
  deriving (Show, Eq)

instance KnownNat w => Arbitrary (ABV w) where
  arbitrary = ABV . BV.mkBV knownNat <$>
    choose (0, maxUnsigned (knownNat :: NatRepr w))

prop_encryptDecrypt :: Eq plaintext
                    => PrivateKeyScheme key plaintext ciphertext
                    -> KeyLength
                    -> plaintext
                    -> Property
prop_encryptDecrypt s (KeyLength n) plaintext = monadicIO $ do
  key <- run $ generateKey s n
  ciphertext <- run $ encrypt s key plaintext
  assert $ decrypt s key ciphertext == plaintext

prop_encryptDecrypt' :: Eq plaintext
                     => PrivateKeyScheme key plaintext ciphertext
                     -> plaintext
                     -> Property
prop_encryptDecrypt' = flip prop_encryptDecrypt (KeyLength undefined)

tests :: TestTree
tests = testGroup "Encrypt/Decrypt"
  [ testProperty "mono-alphabetic shift" $
    prop_encryptDecrypt' monoAlphaShift
  , testProperty "poly-alphabetic shift" $
    prop_encryptDecrypt polyAlphaShift
  , testProperty "mono-alphabetic substitution" $
    prop_encryptDecrypt' monoAlphaSubst
  , testProperty "poly-alphabetic substitution" $
    prop_encryptDecrypt polyAlphaSubst
  , testProperty "one-time pad" $
    prop_encryptDecrypt $
    iso id (prism' unABV (Just . ABV)) id (oneTimePad (knownNat @32))
  ]

main :: IO ()
main = defaultMain tests


src/Crypto/Types.hs
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE TypeFamilies #-}

-- | Miscellaneous types used by various other modules in this library.

module Crypto.Types
  ( -- * Text types
    Alpha(..)
  , alphaFromChar
  , alphasFromString
  , alphaToChar
  , shiftAlpha
  , permuteAlpha
  ) where

import Data.Array.IArray ((!))
import Data.Char (toUpper)
import Data.Maybe (mapMaybe)
import Math.Combinat.Permutations
import Test.QuickCheck

-- | Alphabetical character, for schemes that only use letters.
data Alpha = A | B | C | D | E | F | G | H
           | I | J | K | L | M | N | O | P
           | Q | R | S | T | U | V | W | X
           | Y | Z
  deriving (Show, Eq, Ord, Enum)

instance Arbitrary Alpha where
  arbitrary = chooseEnum (A, Z)

-- | Convert a 'Char' to an 'Alpha'. Throw error if the input is not one of the
-- 26 uppercase letters.
alphaFromChar :: Char -> Maybe Alpha
alphaFromChar 'A' = Just A
alphaFromChar 'B' = Just B
alphaFromChar 'C' = Just C
alphaFromChar 'D' = Just D
alphaFromChar 'E' = Just E
alphaFromChar 'F' = Just F
alphaFromChar 'G' = Just G
alphaFromChar 'H' = Just H
alphaFromChar 'I' = Just I
alphaFromChar 'J' = Just J
alphaFromChar 'K' = Just K
alphaFromChar 'L' = Just L
alphaFromChar 'M' = Just M
alphaFromChar 'N' = Just N
alphaFromChar 'O' = Just O
alphaFromChar 'P' = Just P
alphaFromChar 'Q' = Just Q
alphaFromChar 'R' = Just R
alphaFromChar 'S' = Just S
alphaFromChar 'T' = Just T
alphaFromChar 'U' = Just U
alphaFromChar 'V' = Just V
alphaFromChar 'W' = Just W
alphaFromChar 'X' = Just X
alphaFromChar 'Y' = Just Y
alphaFromChar 'Z' = Just Z
alphaFromChar _ = Nothing

-- | Convert an 'Alpha' to a 'Char'.
alphaToChar :: Alpha -> Char
alphaToChar A = 'A'
alphaToChar B = 'B'
alphaToChar C = 'C'
alphaToChar D = 'D'
alphaToChar E = 'E'
alphaToChar F = 'F'
alphaToChar G = 'G'
alphaToChar H = 'H'
alphaToChar I = 'I'
alphaToChar J = 'J'
alphaToChar K = 'K'
alphaToChar L = 'L'
alphaToChar M = 'M'
alphaToChar N = 'N'
alphaToChar O = 'O'
alphaToChar P = 'P'
alphaToChar Q = 'Q'
alphaToChar R = 'R'
alphaToChar S = 'S'
alphaToChar T = 'T'
alphaToChar U = 'U'
alphaToChar V = 'V'
alphaToChar W = 'W'
alphaToChar X = 'X'
alphaToChar Y = 'Y'
alphaToChar Z = 'Z'

-- | Converts a string to a list of 'Alpha's by mapping the characters to
-- uppercase and removing characters that are not alphabetical.
alphasFromString :: String -> [Alpha]
alphasFromString = mapMaybe (alphaFromChar . toUpper)

-- | Shift an 'Alpha' by a given amount. The input 'Int' does not have to be
-- between @0@ and @25@.
shiftAlpha :: Int -> Alpha -> Alpha
shiftAlpha i = toEnum . (`mod` 26) . (+i) . fromEnum

-- | Apply a permutation to an 'Alpha'. Assumes the input permutation is a
-- permutation on @[0..25]@.
permuteAlpha :: Permutation -> Alpha -> Alpha
permuteAlpha sigma a = fromPermIx (permutationUArray sigma ! toPermIx a)
  where toPermIx = (+ 1) . fromEnum
        fromPermIx = toEnum . subtract 1


src/Crypto/Utils.hs
-- | Helper functions.
module Crypto.Utils
  ( slices
  , unSlices
  , avg
  , sequencePreferred
  -- * Distributions and related functions
  , Distribution
  , getDist
  , NDistribution
  , getNDist
  , englishDist
  , ioc
  , d
  , d'
  -- * Ngrams and related functions
  , NgramCount
  , ngrams
  , ngramsFromFile
  ) where

import Crypto.Types

import Data.List (transpose, permutations, nub, tails)
import Data.List.Split (chunksOf, splitOn)
import qualified Data.Map as Map
import Data.Maybe (mapMaybe)
import Math.Combinat.Compositions

-- | 'slices' splits a list into @n@ pieces. The pieces are constructed by
-- taking every @n+k@th element of the list, where @k@ ranges from @0@ to @n-1@.
-- The last pieces will be shorter if @n@ does not divide the length of the
-- list. If @n <= 0@, @slices n l@ will return an error.
--
-- @
-- slices 3 [0..10] = [[0,3,6,9],[1,4,7,10],[2,5,8]]
-- @
--
-- @slices i@ is the @transpose@ of @chunksOf i@.
slices :: Int -> [a] -> [[a]]
slices i | i <= 0 = error "slices called on non-positive argument"
slices i = transpose . chunksOf i

-- | The inverse of 'slices'.
unSlices :: [[a]] -> [a]
unSlices = concat . transpose

-- | Average of a list of @Fractional a => a@.
avg :: Fractional a => [a] -> a
avg as = sum as / fromIntegral (length as)

-- | Generate all unique permutations of a list.
uniquePermutations :: Eq a => [a] -> [[a]]
uniquePermutations = nub . permutations

-- | Generate indices in preferred order.
--
-- TODO: Document this function better by explaining what "preferred order"
-- means, and come up with a better name for it.
indices :: Int
        -- ^ number of valid indices for each list
        -> Int
        -- ^ number of lists to index into
        -> [[Int]]
indices numRows numCols =
  -- Generate compositions in lexicographic order.
  let comps = reverse <$> compositions numRows numCols
  -- Map each composition to its list of unique permutations, and concatenate the result.
  in concatMap (uniquePermutations . concatMap (uncurry replicate) . flip zip [0..]) comps

-- | Like @sequence@ for a list of lists, but modify the order of the output by
-- using the /preferred order/.
sequencePreferred :: Int -> Int -> [[a]] -> [[a]]
sequencePreferred numRows numCols as =
  let ixss = indices numRows numCols
  in map (zipWith (!!) as) ixss

-- | A distribution of @a@s is a occurrence map, assigning each @a@ to some
-- numeric type @b@.
type Distribution a b = Map.Map a b

-- | Create a simple 'Distribution' by counting occurrences.
getDist :: Ord a => [a] -> Distribution a Integer
getDist as = foldr (flip (Map.insertWith (+)) 1) Map.empty as

-- | A normalized 'Distribution', where the values are normalized to 'Float's
-- that add up to @1@.
type NDistribution a = Distribution a Float

-- | Compute a 'NDistribution' from an input list.
getNDist :: Ord a => [a] -> NDistribution a
getNDist as = fmap ((/ fromIntegral (length as)) . fromInteger) (getDist as)

-- | Average letter frequencies for English-language text, as given in
-- Katz/Lindell page 13, Figure 1.2.
englishDist :: NDistribution Alpha
englishDist = Map.fromList
  [ (A, 0.082)
  , (B, 0.015)
  , (C, 0.028)
  , (D, 0.042)
  , (E, 0.127)
  , (F, 0.022)
  , (G, 0.020)
  , (H, 0.061)
  , (I, 0.070)
  , (J, 0.001)
  , (K, 0.008)
  , (L, 0.040)
  , (M, 0.024)
  , (N, 0.067)
  , (O, 0.075)
  , (P, 0.019)
  , (Q, 0.001)
  , (R, 0.060)
  , (S, 0.063)
  , (T, 0.090)
  , (U, 0.028)
  , (V, 0.010)
  , (W, 0.024)
  , (X, 0.020)
  , (Y, 0.001)
  , (Z, 0.001)
  ]

-- | Given two distributions over the same type, get their index of coincidence.
-- Computes @Sum p_i*q_i@, where @i@ indexes the @a@ type, and @p_i@, @q_i@ are
-- the probabilities at @i@ of the two input distributions.
ioc :: Ord a => NDistribution a -> NDistribution a -> Float
ioc d1 d2 =
  let as = nub (Map.keys d1 ++ Map.keys d2)
  in sum $ map (\a -> Map.findWithDefault 0.0 a d1 * Map.findWithDefault 0.0 a d2) as

-- | The distance between two distributions based on the index of coincidence.
-- This works even if the keys of one of both of the distributions have been
-- permuted. This should be small if the second distribution is /encrypted
-- plaintext/, where the encryption is performed with a simple substitution
-- cipher, assuming the first is the probability distribution we'd expect from
-- plaintext.
d :: Ord a => NDistribution a -> NDistribution a -> Float
d dist1 dist2 = abs (dist1 `ioc` dist1 - dist2 `ioc` dist2)

-- | Computes the distance of the second distribution from the first. This
-- should be small if the second distribution is /decrypted plaintext/, assuming
-- the first is the probability distribution we'd expect from plaintext.
d' :: Ord a => NDistribution a -> NDistribution a -> Float
d' dist1 dist2 = abs (dist1 `ioc` dist1 - dist1 `ioc` dist2)

-- | An 'NgramCount' is just an occurrence count of lists of characters.
type NgramCount a = Distribution [a] Integer

-- | Given @n@ and a text, compute the probability distribution of all @n@-grams
-- in the text.
ngrams :: Ord a => Int -> [a] -> NgramCount a
ngrams n = getDist . mapMaybe (takeMaybe n) . tails

takeMaybe :: Int -> [a] -> Maybe [a]
takeMaybe n _ | n <= 0 = Just []
takeMaybe _ []         = Nothing
takeMaybe n (x:xs)     = (x:) <$> takeMaybe (n-1) xs

-- | Generate an 'NgramCount' from a file in the following format:
--
-- @
-- <ngram> <count>
-- <ngram> <count>
-- ...
-- @
--
-- where each <ngram> is a capitalized n-gram with alphabetical characters and
-- each <count> is a positive integer.
ngramsFromFile :: FilePath -> IO (NgramCount Alpha)
ngramsFromFile path = do
  lns <- lines <$> readFile path
  let f ln = let [ngramStr, countStr] = splitOn " " ln
             in (alphasFromString ngramStr, read countStr)
  return $ Map.fromList (map f lns)


src/Crypto/Attacks.hs
{-# LANGUAGE RankNTypes #-}

-- | A library of attacks against various encryption schemes found in
-- Katz/Lindell.
module Crypto.Attacks
  ( -- * Types of attacks
    CiphertextOnlyAttack
  , KnownPlaintextAttack
  , ChosenPlaintextAttack
  , ChosenCiphertextAttack
    -- * General attacks
  , bruteForce
  , bruteForceWithDist
  , bruteForceEnglish
  , guessPolyAlphaKeyLength
  , breakPoly
    -- * Scheme-specific attacks
  , breakMonoAlphaShift
  , breakMonoAlphaShiftEnglish
  , breakMonoAlphaShiftKnownPlaintext
  , breakPolyAlphaShiftEnglish
  , breakMonoAlphaSubst
  ) where

import Crypto.Schemes
import Crypto.Types
import Crypto.Utils

import Data.List (sortBy, sortOn, (\\))
import qualified Data.List.NonEmpty as LN
import qualified Data.Map as Map
import Data.Maybe (fromJust)
import Data.Ord (Down(..), comparing)
import Math.Combinat.Permutations

-- | A ciphertext-only attack is an attack that only needs the ciphertext to
-- produce a result. (Katz/Lindell pg. 8)
type CiphertextOnlyAttack key plaintext ciphertext = ciphertext -> [(key, plaintext)]

-- | A known-plaintext attack takes a list of plaintext/ciphertext pairs which
-- were all encrypted with the same key, and uses it to decrypt some other
-- ciphertext encrypted by that key. (Katz/Lindell pg. 8)
type KnownPlaintextAttack key plaintext ciphertext =
  [(plaintext, ciphertext)] -> CiphertextOnlyAttack key plaintext ciphertext

-- | A chosen-plaintext attack is an attack where the encryption function can be
-- called arbitrarily, and the key remains fixed. (Katz/Lindell pg. 8)
type ChosenPlaintextAttack key plaintext ciphertext =
  EncryptFn plaintext ciphertext -> CiphertextOnlyAttack key plaintext ciphertext

-- | A chosen-ciphertext attack is an attack where the decryption function can
-- be called arbitrarily, and the key remains fixed. (Katz/Lindell pg. 8)
type ChosenCiphertextAttack key plaintext ciphertext =
  DecryptFn plaintext ciphertext -> CiphertextOnlyAttack key plaintext ciphertext

-- | A brute-force attack can be applied on any encryption scheme. Given a
-- ciphertext to decrypt, the scheme @s@ it was encrypted with, and a list of
-- @key@s to try, simply apply @decrypt s key ciphertext@ for each @key@ in the
-- list, and give a list of all the results along with the @key@s used to
-- generate them.
bruteForce :: [key] -- ^ list of keys to try
           -> PrivateKeyScheme key plaintext ciphertext
           -> CiphertextOnlyAttack key plaintext ciphertext
bruteForce keys s ciphertext = zip keys (flip (decrypt s) ciphertext <$> keys)

-- | A brute-force attack where @plaintext ~ [plainchar]@ for some @Ord
-- plainchar => plainchar@, sorted by closeness to a given letter distribution.
bruteForceWithDist :: Ord plainchar
                   => NDistribution plainchar -- ^ reference distribution
                   -> [key] -- ^ list of keys to try
                   -> PrivateKeyScheme key [plainchar] ciphertext
                   -> CiphertextOnlyAttack key [plainchar] ciphertext
bruteForceWithDist refDist s keys ciphertext =
  let pairs = bruteForce s keys ciphertext
  in sortBy (comparing (d' refDist . getNDist . snd)) pairs

-- | A brute-force attack where @plaintext ~ [Alpha]@, sorted by closeness to
-- English's letter distribution.
bruteForceEnglish :: [key] -- ^ list of keys to try
                  -> PrivateKeyScheme key [Alpha] ciphertext
                  -> CiphertextOnlyAttack key [Alpha] ciphertext
bruteForceEnglish = bruteForceWithDist englishDist

-- | Guess the key length given a ciphertext encoded using a poly-alphabetic
-- substitution cipher, i.e. a cipher defined using 'poly', assuming it
-- was encoded in English. The output is a list of 'Int's, sorted with the
-- likeliest key length at the head of the list, where each 'Int' is paired with
-- the average distance from English's frequency distribution. This should work
-- regardless of the underlying per-character cipher (since the underlying
-- cipher is necessarily a substitution).
guessPolyAlphaKeyLength :: [Int]
                        -- ^ List of key lengths to try
                        -> [Alpha]
                        -> [(Int, Float)]
guessPolyAlphaKeyLength keyLengths ciphertext =
  sortBy (comparing snd) $ zip keyLengths (map f keyLengths)
  where f = avg . map (d englishDist . getNDist) . flip slices ciphertext

-- | Given an attack on a scheme produced by 'mono', lift it to an attack
-- on a scheme produced by 'poly' assuming we know the key length. Use in
-- conjunction with 'guessPolyAlphaKeyLength'.
breakPoly :: CiphertextOnlyAttack key [plainchar] [cipherchar]
          -- ^ Attack on mono cipher
          -> Int
          -- ^ Key length of poly cipher
          -> CiphertextOnlyAttack (LN.NonEmpty key) [plainchar] [cipherchar]
breakPoly attack keyLength ciphertext =
  let sliceAttacks = map attack (slices keyLength ciphertext)
      orderedKeysWithSlices = sequencePreferred 26 keyLength sliceAttacks
      combineKeysWithSlices keysWithSlices =
        (LN.fromList (fst <$> keysWithSlices), unSlices (snd <$> keysWithSlices))
      orderedAttacks = map combineKeysWithSlices orderedKeysWithSlices
  in orderedAttacks

-- | Brute-force, ciphertext-only attack on 'monoAlphaShift'.
breakMonoAlphaShift :: CiphertextOnlyAttack Int [Alpha] [Alpha]
breakMonoAlphaShift = bruteForce [0..25] monoAlphaShift

-- | Brute-force, English-biased ciphertext-only attack on 'monoAlphaShift'.
breakMonoAlphaShiftEnglish :: CiphertextOnlyAttack Int [Alpha] [Alpha]
breakMonoAlphaShiftEnglish = bruteForceEnglish [0..25] monoAlphaShift

-- | English-biased ciphertext-only attack on 'polyAlphaShift', given the
-- key length of the cipher. Use in conjunciton with 'guessPolyAlphaKeyLength'.
breakPolyAlphaShiftEnglish :: Int -> CiphertextOnlyAttack (LN.NonEmpty Int) [Alpha] [Alpha]
breakPolyAlphaShiftEnglish = breakPoly breakMonoAlphaShiftEnglish

-- | Known-plaintext attack on 'monoAlphaShift'. Assumes the input pairs are valid;
-- does not check for this. If the input list contains only pairs of empty
-- strings, this degenerates into a brute-force attack. Assuming any of the
-- input pairs contain nonempty strings, this is guaranteed to be correct for
-- any shift cipher, and only produces one result.
breakMonoAlphaShiftKnownPlaintext :: KnownPlaintextAttack Int [Alpha] [Alpha]
breakMonoAlphaShiftKnownPlaintext ((p:_,c:_):_) ciphertext =
  let key = (fromEnum c - fromEnum p) `mod` 26
  in [(key, decrypt monoAlphaShift key ciphertext)]
breakMonoAlphaShiftKnownPlaintext (_:pairs) ciphertext =
  breakMonoAlphaShiftKnownPlaintext pairs ciphertext
breakMonoAlphaShiftKnownPlaintext [] ciphertext = breakMonoAlphaShift ciphertext

-- | Attempts to break a mono-alphabetic substitution cipher given an expected
-- distribution of alphabetical characters. Needs improvement.
breakMonoAlphaSubst :: NDistribution Alpha
                    -> CiphertextOnlyAttack Permutation [Alpha] [Alpha]
breakMonoAlphaSubst refDist ciphertext =
  let dist = getNDist ciphertext
      distLetters' = fst <$> sortOn (Down . snd) (Map.toList dist)
      distLetters = distLetters' ++ ([A .. Z] \\ distLetters')
      refLetters'  = fst <$> sortOn (Down . snd) (Map.toList refDist)
      refLetters = refLetters' ++ ([A .. Z] \\ refLetters')
      pairs = zip refLetters distLetters
      sigma = toPermutation $
        map (\a -> 1 + fromEnum (fromJust (lookup a pairs))) [A .. Z]
  in [(sigma, decrypt monoAlphaSubst sigma ciphertext)]


src/Crypto/Schemes.hs
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeOperators #-}

-- | Types for, and examples of, encryption schemes found in Katz/Lindell.
module Crypto.Schemes
  ( -- * Encryption and decryption
    EncryptFn
  , DecryptFn
    -- * Private key encryption
    -- ** Private key encryption schemes
  , PrivateKeyScheme(..)
  , generateKey'
    -- ** New schemes from old
  , iso
  , mono
  , poly
    -- ** Example private key ciphers
  , alphaShift
  , monoAlphaShift
  , polyAlphaShift
  , alphaSubst
  , monoAlphaSubst
  , polyAlphaSubst
  , oneTimePad
  ) where

import Crypto.Types

import Control.Lens (Iso', Prism', from, (^.), re, (^?))
import Control.Monad.Random
import qualified Data.BitVector.Sized as BV
import Data.Foldable (toList)
import qualified Data.List.NonEmpty as LN
import Data.Maybe (fromJust)
import Math.Combinat.Permutations

-- | Encryption function mapping @plaintext@ to @ciphertext@. Note that
-- encryption can be probabilistic (but doesn't have to be).
type EncryptFn plaintext ciphertext =
  forall m . MonadRandom m => plaintext -> m ciphertext

-- | Decryption function mapping @ciphertext@ to @plaintext@. Note that
-- decryption is deterministic.
type DecryptFn plaintext ciphertext = ciphertext -> plaintext

-- | Private key scheme, as defined in Katz/Lindell page 60 (Definition 3.7),
-- but with generalized keys, plaintext, and ciphertext types.
--
-- Note that, for scheme @s@ to be valid, we require that for all keys @k@ and
-- messages @p@:
--
-- @
--   decrypt s k (encrypt s k m) == m
-- @
data PrivateKeyScheme key plaintext ciphertext = PrivateKeyScheme
  { generateKey :: forall m . MonadRandom m => Int -> m key
    -- ^ Generate a random key from a given security parameter ('Int').
  , encrypt :: key -> EncryptFn plaintext ciphertext
    -- ^ Encrypt plaintext with a given key.
  , decrypt :: key -> DecryptFn plaintext ciphertext
    -- ^ Decrypt plaintext with a given key.
  }

-- | Generate a key with a security parameter of @undefined@. This is useful for
-- schemes that are known to ignore the security parameter (like a basic shift
-- or substitution cipher).
generateKey' :: forall key plaintext ciphertext m . MonadRandom m
             => PrivateKeyScheme key plaintext ciphertext -> m key
generateKey' = flip generateKey undefined

-- | Generate a 'PrivateKeyScheme' from an existing scheme by supplying
-- bidirectional mappings between the @key@, @plaintext@, and @ciphertext@
-- types. bijections between the key, plaintext, and ciphertext types.
--
-- Note the argument types:
-- * @Iso' key key'@, an isomorphism between @key@ and @key'@
-- * @Prism' plaintext plaintext'@, a reversible injective embedding of
--   @plaintext'@ into @plaintext@
-- * @Prism' ciphertext' ciphertext@, a reversible injective embedding of
--   @ciphertext@ into @ciphertext'
iso :: Iso' key key'
    -> Prism' plaintext plaintext'
    -> Prism' ciphertext' ciphertext
    -> PrivateKeyScheme key plaintext ciphertext
    -> PrivateKeyScheme key' plaintext' ciphertext'
iso kl pl cl s = PrivateKeyScheme
  { generateKey = \n -> do
      key <- generateKey s n
      return $ key ^. kl
  , encrypt = \key' plaintext' -> do
      ciphertext <- encrypt s (key' ^. from kl) (plaintext' ^. re pl)
      return $ ciphertext ^. re cl
  , decrypt = \key' ciphertext' ->
      let plaintext = decrypt s (key' ^. from kl) (fromJust $ ciphertext' ^? cl)
      in fromJust $ plaintext ^? pl
  }

-- | Given a 'PrivateKeyScheme' that operates on individual characters, lift
-- that scheme to one that operates on strings. The new scheme uses the same
-- @key@ type as the per-character scheme, and encrypts/decrypts by mapping the
-- input scheme's 'encrypt' and 'decrypt' functions over the strings.
mono :: PrivateKeyScheme key plainchar cipherchar
     -> PrivateKeyScheme key [plainchar] [cipherchar]
mono s = PrivateKeyScheme
  { generateKey = generateKey s
  , encrypt = traverse . encrypt s
  , decrypt = map . decrypt s
  }

-- | Given a 'PrivateKeyScheme' that operates on individual characters, generate
-- a new scheme that operates on strings. The new scheme uses a /list/ of
-- @key@s, and encrypts/decrypts by cycling through this list, applying each
-- @key@ to each character of plaintext. When we run out of keys, start over
-- with the original list.
--
-- The security parameter of the lifted scheme will determine the length of the
-- key produced by the 'generateKey' function. The 'Int' that is passed to this
-- function will be used as the security parameter that gets fed to the
-- 'generateKey' of the input scheme.
--
-- If the key length is non-positive, the key generation will throw a runtime
-- error.
poly :: PrivateKeyScheme key plaintext ciphertext
     -> Int
     -- ^ The security parameter to pass to the input scheme's
     -- key generator.
     -> PrivateKeyScheme (LN.NonEmpty key) [plaintext] [ciphertext]
poly s securityParam = PrivateKeyScheme
  { generateKey = \keyLength -> do
      ks <- replicateM keyLength (generateKey s securityParam)
      case LN.nonEmpty ks of
        Just key -> return key
        Nothing -> error msg
  , encrypt = zipWithM (encrypt s) . cycle . toList
  , decrypt = zipWith (decrypt s) . cycle . toList
  }

  where msg = "generateKey called with non-positive key length"

-- | Shift cipher for single 'Alpha'. This is used to define
-- 'monoAlphaShift' and 'polyAlphaShift'. The key generator ignores
-- its input.
alphaShift :: PrivateKeyScheme Int Alpha Alpha
alphaShift = PrivateKeyScheme
  { generateKey = const $ uniform [0 .. 25]
  , encrypt = \key -> return . shiftAlpha key
  , decrypt = shiftAlpha . negate
  }

-- | Mono-alphabetic shift cipher. The key is an 'Int' between 0 and 25
-- (inclusive), and we shift each character by that amount to encrypt.
--
-- @
-- monoAlphaShift == mono alphaShift
-- @
monoAlphaShift :: PrivateKeyScheme Int [Alpha] [Alpha]
monoAlphaShift = mono alphaShift

-- | Poly-alphabetic shift cipher, also known as Vigenère cipher. This promotes
-- the normal 'alphaShift' to operate on lists of keys.
--
-- @
-- polyAlphaShift == poly alphaShift undefined
-- @
polyAlphaShift :: PrivateKeyScheme (LN.NonEmpty Int) [Alpha] [Alpha]
polyAlphaShift = poly alphaShift undefined

-- | Substitution cipher for single 'Alpha'. This is used to define
-- 'monoAlphaSubst' and 'polyAlphaSubst'. The key generator ignores
-- its input.
alphaSubst :: PrivateKeyScheme Permutation Alpha Alpha
alphaSubst = PrivateKeyScheme
  { generateKey = const $ fst . randomPermutation 26 . mkStdGen <$> getRandom
  , encrypt = \key -> return . permuteAlpha key
  , decrypt = permuteAlpha . inverse
  }

-- | Mono-alphabetic substitution cipher. The key is a 'Permutation' of the
-- alphabet, and we apply the permutation to each character in the input to
-- encrypt.
--
-- @
-- monoAlphaSubst == mono alphaSubst
-- @
monoAlphaSubst :: PrivateKeyScheme Permutation [Alpha] [Alpha]
monoAlphaSubst = mono alphaSubst

-- | Poly-alphabetic substitution cipher. This promotes the normal
-- 'alphaSubst' to operate on lists of keys.
--
-- @
-- polyAlphaSubst == poly alphaSubst undefined
-- @
polyAlphaSubst :: PrivateKeyScheme (LN.NonEmpty Permutation) [Alpha] [Alpha]
polyAlphaSubst = poly alphaSubst undefined

-- | One-time pad on bitvectors of a given length.
oneTimePad :: BV.NatRepr w -> PrivateKeyScheme (BV.BV w) (BV.BV w) (BV.BV w)
oneTimePad w = PrivateKeyScheme
  { generateKey = const $ BV.mkBV w <$> getRandom
  , encrypt = \key -> return . BV.xor key
  , decrypt = BV.xor
  }
